core/TypeErrorDiagnostics.cc

```diff
@@ -55,19 +55,46 @@ void TypeErrorDiagnostics::maybeAutocorrect(const GlobalState &gs, ErrorBuilder
     if (gs.suggestUnsafe.has_value()) {
         e.replaceWith(fmt::format("Wrap in `{}`", *gs.suggestUnsafe), loc, "{}({})", *gs.suggestUnsafe,
                       loc.source(gs).value());
-    } else {
-        auto withoutNil = Types::approximateSubtract(gs, actualType, Types::nilClass());
-        if (!withoutNil.isBottom() &&
-            Types::isSubTypeUnderConstraint(gs, constr, withoutNil, expectedType, UntypedMode::AlwaysCompatible)) {
-            e.replaceWith("Wrap in `T.must`", loc, "T.must({})", loc.source(gs).value());
-        } else if (Types::isSubTypeUnderConstraint(gs, constr, expectedType, Types::Boolean(),
-                                                   UntypedMode::AlwaysCompatible)) {
-            if (core::isa_type<ClassType>(actualType)) {
-                auto classSymbol = core::cast_type_nonnull<ClassType>(actualType).symbol;
-                if (classSymbol.exists() && classSymbol.data(gs)->owner == core::Symbols::root() &&
-                    classSymbol.data(gs)->name == core::Names::Constants::Boolean()) {
-                    e.replaceWith("Prepend `!!`", loc, "!!({})", loc.source(gs).value());
-                }
+        return;
+    }
+
+    // If we expected an array of T, but got an array of nilable T, add an autocorrect
+    // to add .compact.
+    auto arrayOfUntyped = Types::arrayOfUntyped();
```

Comment:
I try not to use untyped in normal (non-error recovery code) if I can avoid it. I probably would have written this to do the two cast operations first, then call derivesFrom Symbols::Array() on the klass field in the appliedtype. This also has the side benefit of more predictable performance (the cast check is much faster than an arbitrary call to check subtyping)